### ISSUE TITLE
memory leak in torch>=2.0.0

### ISSUE THREAD

**Issue by HuangdaoQin** [pytorch/pytorch#108334]
### 🐛 Describe the bug

When I execute the following code in torch>=2.0.0 torchvision>=0.15.1, cpu memory leaks; when I execute the following code in torch==0.13.0 torchvision==0.14.0, cpu memory does not leak. The memory leak is worse if I perform more tensor operations in getitem and collate_fn.

```python
import random
import torch
from torch.utils.data import Dataset, DataLoader
import torchvision
import psutil
import os

class ExampleDataset(Dataset):
    def __init__(self):
        self.num_samples = 1000

    def __len__(self):
        return self.num_samples

    def __getitem__(self, idx):
        h = random.randint(666, 999)
        w = random.randint(666, 999)
        tensor = torch.ones((3, h, w))
        return tensor
    
dataset = ExampleDataset()
data_loader = DataLoader(dataset, batch_size=1, num_workers=0)
model = getattr(torchvision.models, 'resnet18')().to('cuda')

for epoch in range(10):
    for tensor in data_loader:
        tensor = tensor.to('cuda')
        output = model(tensor)
        print('memory used:', psutil.Process(os.getpid()).memory_info().rss / 1024.0 / 1024.0)
```

### Versions

Package                  Version
------------------------ ----------
certifi                  2023.7.22
charset-normalizer       3.2.0
cmake                    3.27.2
filelock                 3.12.3
idna                     3.4
Jinja2                   3.1.2
lit                      16.0.6
MarkupSafe               2.1.3
mpmath                   1.3.0
networkx                 3.1
numpy                    1.25.2
nvidia-cublas-cu11       11.10.3.66
nvidia-cuda-cupti-cu11   11.7.101
nvidia-cuda-nvrtc-cu11   11.7.99
nvidia-cuda-runtime-cu11 11.7.99
nvidia-cudnn-cu11        8.5.0.96
nvidia-cufft-cu11        10.9.0.58
nvidia-curand-cu11       10.2.10.91
nvidia-cusolver-cu11     11.4.0.1
nvidia-cusparse-cu11     11.7.4.91
nvidia-nccl-cu11         2.14.3
nvidia-nvtx-cu11         11.7.91
Pillow                   10.0.0
pip                      23.2.1
psutil                   5.9.5
requests                 2.31.0
setuptools               68.0.0
sympy                    1.12
torch                    2.0.0
torchvision              0.15.1
triton                   2.0.0
typing_extensions        4.7.1
urllib3                  2.0.4
wheel                    0.38.4

**Comment by HuangdaoQin** [comment]
@ptrblck thx~

**Comment by tringwald** [comment]
I found an even simpler example that reproduces the leak:

```python
import random
import torch
import psutil
import os

model = torch.nn.Conv2d(in_channels=3, out_channels=2000, kernel_size=3, stride=1, padding=1).cuda()
leak_memory = True
h = w = 224

for epoch in range(100000):
    if leak_memory:
        h = random.randint(5, 800)
        w = random.randint(5, 800)
    tensor = torch.ones((1, 3, h, w)).cuda()
    output = model(tensor)
    print('\rmemory used:', psutil.Process(os.getpid()).memory_info().rss / 1024.0 / 1024.0, end='')

```

It seems related to nn.Conv2d, but only when dynamically changing the spatial dimensions of the input shape. With `leak_memory=False`, the input size remains constant at (1, 3, 224, 224) and the memory usage remains constant too.

**Comment by tringwald** [comment]
Seems to be a duplicate of #98688, which is already fixed. The example above doesn't leak memory anymore on the current nightly build `2.1.0.dev20230831+cu118`.

**Comment by HuangdaoQin** [comment]
> as others have pointed out this was fixed already

this issue still occur in torch 2.1.0

**Comment by tringwald** [comment]
For me, neither the initial code nor my own example above leak memory on `2.1.0`. Memory usage only rises until the execution plan LRU cache is filled. After that, it stays constant. If the default limit of 10,000 entries is too large for you, you might want to set `TORCH_CUDNN_V8_API_LRU_CACHE_LIMIT`.
